File: src/test/java/javaKinds/BasicPolymorphismTest.java
package javaKinds;

import java.util.ArrayList;
import java.util.List;

import org.junit.Test;

public class BasicPolymorphismTest {

	@Test
	public void itShouldMakeAnimalSounds() {
		List<Animal> animals = new ArrayList<Animal>();
		animals.add(new Dog());
		animals.add(new Cat());
		animals.add(new Donkey());

		for (Animal animal : animals) {
			System.out.println(animal.speak());
		}

		/*
		 * Output:
		 * 
		 * woof 
		 * mew mew 
		 * hee-haw
		 */
	}

	class Zerg extends Animal {

		@Override
		public String speak() {
			return "for the swarm";
		}
	}

	@Test
	public void itShouldAllowYouToDefineNewAnimals() {
		List<Animal> animals = new ArrayList<Animal>();
		animals.add(new Dog());
		animals.add(new Cat());
		animals.add(new Donkey());
		animals.add(new Zerg());

		for (Animal animal : animals) {
			System.out.println(animal.speak());
		}
		/*
		 * Output:
		 * 
		 * woof 
		 * mew mew 
		 * hee-haw
		 * for the swarm
		 */
	}
}

File: FamilyTies/src/test/scala/scalaKinds/BasicPolymorphismTest.scala
package scalaKinds
import org.scalatest._

class BasicPolymorphismTest extends FunSpec with ShouldMatchers {
  it("should make animal sounds") {
    val animals = Seq(Dog(), Cat(), Donkey())
    for (animal <- animals) println(animal.speak())
    /*
		 * Output:
		 * 
		 * woof 
		 * mew mew 
		 * hee-haw
		 */
  }

  it("should allow you to define new animals") {
    
    case class Zerg() extends Animal(speak = () => "for the swarm")
    
    val animals = Seq(Dog(), Cat(), Donkey(), Zerg())
    for (animal <- animals) println(animal.speak())
    /*
		 * Output:
		 * 
		 * woof 
		 * mew mew 
		 * hee-haw
		 * for the swarm
		 */
  }
}
